### ISSUE TITLE
For min/max functions a default value changes the key function type

### ISSUE THREAD

**Issue by kaste** [python/mypy#6460]
Related but maybe not the same #3354

Given

```
from typing import List, Tuple
l: List[Tuple[int, str]] = []
v = max(l, key=lambda item: item[0], default=None)
reveal_type(v)
```

mypy complains about the `item[0]` usage in the key function:

```
Value of type "Optional[Tuple[int, str]]" is not indexable
Revealed type is 'Union[Tuple[builtins.int, builtins.str], None]'
```

Also interesting that it shows two different representations for the inferred type. 
For user ergonomics the revealed type should be 'Optional[Tuple[int, str]]' as well, no?

**Comment by gvanrossum** [comment]
Maybe should this be closed here and opened in typeshed instead? @srittau. 

Agreed on using the same representation for types; can you file that separately? Or send in a PR if you can.

**Comment by kaste** [comment]
Well, changing `reveal_type` is probably easy but then we get approx. 1.000.000 test failures. E.g.

```
Expected:
  main:10: error: Revealed type is 'builtins.int' (diff)
  main:11: error: Revealed type is 'Tuple[builtins.int, builtins.int]' (diff)
  main:13: error: Revealed type is 'Literal[1]' (diff)
  main:14: error: Argument 1 to "force2" has incompatible type "Tuple[int, in...
  main:14: error: Revealed type is 'Tuple[builtins.int, builtins.int]' (diff)
Actual:
  main:10: error: Revealed type is "int"        (diff)
  main:11: error: Revealed type is "Tuple[int, int]" (diff)
  main:13: error: Revealed type is "Literal[1]" (diff)
  main:14: error: Argument 1 to "force2" has incompatible type "Tuple[int, in...
  main:14: error: Revealed type is "Tuple[int, int]" (diff)
```

Say, I find my regex-ninja in me, the diff of the PR would be 💣.

**Comment by srittau** [comment]
Original issue was fixed in python/typeshed#2833.

**Comment by kaste** [comment]
I think a 8 lines diff still counts as contribution to python core. 🙌
